wrappers/services.go

```diff
@@ -59,25 +59,55 @@ func generateSnapServiceFile(app *snap.AppInfo) (string, error) {
 }
 
 // StartSnapServices starts service units for the applications from the snap which are services.
-func StartSnapServices(s *snap.Info, inter interacter) error {
+func StartSnapServices(s *snap.Info, inter interacter) (err error) {
+	hasReloaded := false
+	sysd := systemd.New(dirs.GlobalRootDir, inter)
+	var enabled []*snap.AppInfo
+	var started []*snap.AppInfo
+
+	defer func() {
```

Comment:
I have a non-binding suggestion for your consideration

```
for stuff := range workList {
  err = do(stuff)
  if err {
    return err
  }
  defer undoOnError(stuff)
}
```

That is, the idea is to follow the open/defer close` model but have the defer _only_ do cleanup if an error really occurred. This would have the (subjective) advantage that the logic may be shorter and more closely placed. The code you wrote is fine but one has to keep the defer helper in sync with what happens below.